Description: add tests for fillMediaBag/extractMedia
Origin: upstream, https://github.com/jgm/pandoc/commit/fe62da6
Author: John MacFarlane <jgm@berkeley.edu>
Bug: https://github.com/jgm/pandoc/security/advisories/GHSA-xj5q-fv23-575g
Bug-Debian: https://security-tracker.debian.org/tracker/CVE-2023-35936
Forwarded: yes
Last-Update: 2023-07-25
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/pandoc.cabal
+++ b/pandoc.cabal
@@ -791,6 +791,7 @@
                   tasty-lua         >= 1.0     && < 1.1,
                   tasty-quickcheck  >= 0.8     && < 0.11,
                   text              >= 1.1.1.0 && < 2.1,
+                  temporary         >= 1.1     && < 1.4,
                   time              >= 1.5     && < 1.14,
                   xml               >= 1.3.12  && < 1.4,
                   zip-archive       >= 0.2.3.4 && < 0.5
@@ -800,6 +801,7 @@
                   Tests.Lua
                   Tests.Lua.Module
                   Tests.Shared
+                  Tests.MediaBag
                   Tests.Readers.LaTeX
                   Tests.Readers.HTML
                   Tests.Readers.JATS
--- /dev/null
+++ b/test/Tests/MediaBag.hs
@@ -0,0 +1,39 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Tests.MediaBag (tests) where
+
+import Test.Tasty
+import Test.Tasty.HUnit
+-- import Tests.Helpers
+import Text.Pandoc.Class (extractMedia, fillMediaBag, runIOorExplode)
+import System.IO.Temp (withTempDirectory)
+import System.FilePath
+import Text.Pandoc.Builder as B
+import System.Directory (doesFileExist, copyFile, setCurrentDirectory, getCurrentDirectory)
+
+tests :: [TestTree]
+tests = [
+  testCase "test fillMediaBag & extractMedia" $
+      withTempDirectory "." "extractMediaTest" $ \tmpdir -> do
+        olddir <- getCurrentDirectory
+        setCurrentDirectory tmpdir
+        copyFile "../../test/lalune.jpg" "moon.jpg"
+        let d = B.doc $
+                  B.para (B.image "../../test/lalune.jpg" "" mempty) <>
+                  B.para (B.image "moon.jpg" "" mempty) <>
+                  B.para (B.image "data://image/png;base64,cHJpbnQgImhlbGxvIgo=;.lua+%2f%2e%2e%2f%2e%2e%2fa%2elua" "" mempty) <>
+                  B.para (B.image "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" "" mempty)
+        runIOorExplode $ do
+          fillMediaBag d
+          extractMedia "foo" d
+        exists1 <- doesFileExist ("foo" </> "moon.jpg")
+        assertBool "file in directory extract with original name" exists1
+        exists2 <- doesFileExist ("foo" </> "f9d88c3dbe18f6a7f5670e994a947d51216cdf0e.jpg")
+        assertBool "file above directory extracted with hashed name" exists2
+        exists3 <- doesFileExist ("foo" </> "2a0eaa89f43fada3e6c577beea4f2f8f53ab6a1d.lua")
+        exists4 <- doesFileExist "a.lua"
+        assertBool "data uri with malicious payload does not get written to arbitrary location"
+          (exists3 && not exists4)
+        exists5 <- doesFileExist ("foo" </> "d5fceb6532643d0d84ffe09c40c481ecdf59e15a.gif")
+        assertBool "data uri with gif is properly decoded" exists5
+        setCurrentDirectory olddir
+  ]
--- a/test/test-pandoc.hs
+++ b/test/test-pandoc.hs
@@ -51,6 +51,7 @@
 import qualified Tests.Writers.AnnotatedTable
 import qualified Tests.Writers.TEI
 import qualified Tests.Writers.Markua
+import qualified Tests.MediaBag
 import Text.Pandoc.Shared (inDirectory)
 
 tests :: FilePath -> TestTree
@@ -58,6 +59,7 @@
         [ Tests.Command.tests
         , testGroup "Old" (Tests.Old.tests pandocPath)
         , testGroup "Shared" Tests.Shared.tests
+        , testGroup "MediaBag" Tests.MediaBag.tests
         , testGroup "Writers"
           [ testGroup "Native" Tests.Writers.Native.tests
           , testGroup "ConTeXt" Tests.Writers.ConTeXt.tests
